src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Tds/Tokens/DataClassification/DataClassificationTokenParser.cs

```diff
@@ -0,0 +1,115 @@
+
+namespace Microsoft.Data.SqlClientX.Tds.Tokens.DataClassification
+{
+    internal class DataClassificationTokenParser : TokenParser
+    {
+        public override async ValueTask<Token> ParseAsync(TokenType tokenType, TdsStream tdsStream, bool isAsync, CancellationToken ct)
+        {
+            if (tdsStream.DataClassificationVersion == 0)
+            {
+                throw SQL.ParsingError(ParsingErrorState.DataClassificationNotExpected);
+            }
+
+            // get the labels
+            ushort numLabels = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+
+            List<Label> labels = new List<Label>(numLabels);
+            for (ushort i = 0; i < numLabels; i++)
+            {
+                var labelLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var label = await tdsStream.TdsReader.ReadStringAsync(labelLen, isAsync, ct).ConfigureAwait(false);
+                var idLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var id = await tdsStream.TdsReader.ReadStringAsync(idLen, isAsync, ct).ConfigureAwait(false);
+                labels.Add(new Label(label, id));
+            }
+
+            // get the information types
+            ushort numInformationTypes = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct);
+
+            List<InformationType> informationTypes = new List<InformationType>(numInformationTypes);
+            for (ushort i = 0; i < numInformationTypes; i++)
+            {
+                var informationTypeLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var informationType = await tdsStream.TdsReader.ReadStringAsync(informationTypeLen, isAsync, ct).ConfigureAwait(false);
+                var idLen = await tdsStream.TdsReader.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+                var id = await tdsStream.TdsReader.ReadStringAsync(idLen, isAsync, ct).ConfigureAwait(false);
+                informationTypes.Add(new InformationType(informationType, id));
+            }
+
+            // get sensitivity rank
+            int sensitivityRank = (int)SensitivityRank.NOT_DEFINED;
+            if (tdsStream.DataClassificationVersion > TdsEnums.DATA_CLASSIFICATION_VERSION_WITHOUT_RANK_SUPPORT)
+            {
+                var rank = await tdsStream.TdsReader.ReadInt32Async(isAsync, ct).ConfigureAwait(false);
+                if (!Enum.IsDefined(typeof(SensitivityRank), rank))
+                {
+                    return null;
+                }
+            }
+
+            // get the per column classification data (corresponds to order of output columns for query)
+            var numResultColumns = await tdsStream.TdsReader.ReadUInt16Async(isAsync, ct).ConfigureAwait(false);
+            List<ColumnSensitivity> columnSensitivities = new List<ColumnSensitivity>(numResultColumns);
```

Comment:
We're not changing public signature, so this will stay.

Reply:
What public signatures? The token and token parser classes are all internal and implementation details of how tokens are parsed.

Reply:
Ok, I see what you mean by having ReadOnlyCollection as part of the interface, but ReadOnlyCollection can be initialized from an array (since it implements IList)